Python/symtable.c

```diff
@@ -1527,9 +1533,53 @@ symtable_handle_comprehension(struct symtable *st, expr_ty e,
     if (value)
         VISIT_IN_BLOCK(st, expr, value, (void*)e);
     VISIT_IN_BLOCK(st, expr, elt, (void*)e);
+    if (Py_Py3kWarningFlag && st->st_cur->ste_generator) {
+        const char *msg = (
+            (e->kind == SetComp_kind) ? "'yield' inside set comprehension" :
+            (e->kind == DictComp_kind) ? "'yield' inside dict comprehension" :
+            "'yield' inside generator expression");
+        if (!symtable_warn(st, PyExc_DeprecationWarning, msg, -1)) {
+            symtable_exit_block(st, (void *)e);
+            return 0;
+        }
+    }
+    st->st_cur->ste_generator |= is_generator;
     return symtable_exit_block(st, (void *)e);
 }
 
+static int
+symtable_visit_listcomp(struct symtable *st, expr_ty e)
+{
+    asdl_seq *generators = e->v.ListComp.generators;
+    int i;
+    int is_generator = st->st_cur->ste_generator;
```

Comment:
I missed the addition of `is_generator` on my first read-through of this function, so I think it could use an explanatory comment. Something like:

    /* In order to check for yield expressions under '-3', we clear the generator flag, and restore it at the end */